src/Microsoft.OData.Core/ODataBatchWriter.cs

```diff
@@ -514,62 +417,31 @@ private void VerifyCanWriteEndChangeset(bool synchronousCall)
         }
 
         /// <summary>
-        /// Ends an active changeset - implementation of the actual functionality.
+        /// Ends an active changeset.
         /// </summary>
-        private void WriteEndChangesetImplementation()
-        {
-            // write pending message data (headers, response line) for a previously unclosed message/request
-            this.WritePendingMessageData(true);
+        protected abstract void WriteEndChangesetImplementation();
 
-            string currentChangeSetBoundary = this.changeSetBoundary;
-
-            // change the state first so we validate the change set boundary before attempting to write it.
-            this.SetState(BatchWriterState.ChangeSetCompleted);
-
-            // In the case of an empty changeset the start changeset boundary has not been written yet
-            // we will leave it like that, since we want the empty changeset to be represented only as
-            // the end changeset boundary.
-            // Due to WCF DS V2 compatiblity we must not write the start boundary in this case
-            // otherwise WCF DS V2 won't be able to read it (it fails on the start-end boundary empty changeset).
-
-            // write the end boundary for the change set
-            ODataBatchWriterUtils.WriteEndBoundary(this.rawOutputContext.TextWriter, currentChangeSetBoundary, !this.changesetStartBoundaryWritten);

```

Comment:
As above, End*Implementation methods should be private in the base class, and contain shared logic, then call abstract End* methods that are implemented in the derived classes and do the actual format-specific writing.